Adds WPAndroid Integration document (#3262)
wordpress-mobile/gutenberg-mobile: WPAndroid Integration.md

> ### Summary

Reviewer: I think this could benefit from an `h1` title:

```suggestion
# WordPress Android Integration Guide

### Summary
```

When looking at the doc from GitHub, it's not immediately clear what it is about because the file name is as prominent as the `h3` in the rendered markdown:

![image](https://user-images.githubusercontent.com/1218433/111249766-540c3280-8660-11eb-921e-025f1b3077b9.png)

Author: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

> WPAndroid by default integrates `react-native-bridge` as a binary dependency that's fetched from a remote maven repo. The CI in this repo will deploy a new version each time a commit is pushed to `develop` or an open PR, as well as each time a tag is created. This is the only required dependency. Whenever a new version is deployed, we can simply update `ext.gutenbergMobileVersion` in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle).

Reviewer: Totally out of scope, but I thought I'd mention it here. 

For a while now, I've been writing Markdown using a simplified version of [_Semantic Line Breaks_](https://sembr.org/): add a line break each ~~substantial unit of thought~~ sentence.

```suggestion
WPAndroid by default integrates `react-native-bridge` as a binary dependency that's fetched from a remote maven repo.
The CI in this repo will deploy a new version each time a commit is pushed to `develop` or an open PR, as well as each time a tag is created.
This is the only required dependency. Whenever a new version is deployed, we can simply update `ext.gutenbergMobileVersion` in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle).
```

From a code review and source code management point of view, semantic line breaks make it easier to suggest and identify changes, because they reduce the scope from the paragraph to the line.

Author: TIL. Added line breaks to the whole document in 56cd0065eebe143c6034380851f4f4dbd604f4d8.

- [Work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
- [Deploy a new version of `react-native-bridge` from CI](#deploy-a-new-version-of-react-native-bridge-from-ci)
- [Manually deploy a new version of `react-native-bridge`](#manually-deploy-a-new-version-of-react-native-bridge)
- [Update WPAndroid's binary version](#update-wpandroids-binary-version)
- [How to](#how-to)
    - [How to test a gutenberg-mobile PR in WPAndroid](how-to-test-a-gutenberg-mobile-pr-in-wpandroid)
    - [How to share WIP WPAndroid APK](how-to-share-wip-wpandroid-apk)
    - [How to checkout gutenberg-mobile to deployed binary version](#how-to-checkout-gutenberg-mobile-to-deployed-binary-version)

---

### Work with local gutenberg-mobile checkout

> - Make sure you run `npm install` or `npm ci`

Reviewer: I think it would be clearer to suggest only one option here.

I think `npm ci` would be best here, because we're interested in _consuming_ the local checkout more than modifying its libraries versions (something that could happen with `npm install`).

Author: Although I agree with you, the reason I wrote it this way is because `npm install` is the default suggestion everywhere in `gutenberg-mobile` and `gutenberg` repos, at least as far as I can tell. Going against that in this document didn't feel right to me.

I'll leave this decision to @mchowning.

Reviewer 2: I would either leave it as-is or just have `npm install` since that is the default suggestion elsewhere.

> we're interested in consuming the local checkout more than modifying its libraries versions

Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?

Author: I am going to leave it as is. Since the CI does `npm ci`, I think it's good to have that as an alternative.

> Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?

Yes they would, that's the whole idea.

Reviewer: Cool 👍 

> > Is this necessarily true? Won't some devs be using the local gutenberg-mobile checkouts for development?
>
> Yes they would, that's the whole idea.

Out of curiosity, would that entail much work around the Node modules?

Reviewer 2: > Out of curiosity, would that entail much work around the Node modules?

Not the vast majority of the time, but it could.

- Copy [local-builds.gradle-example](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/local-builds.gradle-example) renaming it to `local-builds.gradle`

> - Update `localGutenbergMobilePath`

Reviewer: Even though it may seem obvious, I'd add:

```suggestion
- Update `localGutenbergMobilePath` with the path to gutenberg-mobile in your machine
```

Author: I've opted to the following instead:

> - Update `localGutenbergMobilePath`to your local `gutenberg-mobile`checkout

Simply because I used `local checkout` terminology everywhere else in the document. 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1

Reviewer: 👍

- Run metro server with `npm run start:reset`

> **How it works:**

Reviewer: Nitpick, since this is not a standalone sentence:

```suggestion
#### How it works
```

Author: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

Author: I realized I didn't do this for the other 2 instances, so updated in ba88fd37a9778222b56f2ed7669286cba9150619.

When [settings.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/settings.gradle) finds `localGutenbergMobilePath` in `local-builds.gradle`, it'll substitute the below binary dependency with the local folder:

```
implementation "$rootProject.gradle.ext.gutenbergMobileBinaryPath:$rootProject.ext.gutenbergMobileVersion"
```

> `react-native-bridge` project is setup to include all the other projects, so no other action is necessary.

Reviewer: Nitpick

```suggestion
The `react-native-bridge` project is setup to include all the other projects, so no other action is necessary.
```

or

```suggestion
`react-native-bridge` is setup to include all the other projects, so no other action is necessary.
```

The rationale, possibly flawed, behind my suggestion is that "`react-native-bridge`" is the "first name" of the of the sentence's subject while "project" is a specification of what `react-native-bridge` is (_I think..._)

For example, "Gio" is my name and I'm a reviewer of this PR. "Gio reviewer makes nitpicky comments" doesn't sound as right as "Gio makes nitpicky comments" or "The reviewer Gio makes nitpicky comments". 

**This is an interesting discussion on grammar/writing but either suggestion provides a small epsilon improvement to the document, so 🤷‍♂️**

Author: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

---

### Deploy a new version of `react-native-bridge` from CI

There are a few different ways to do this:

1. Open a new PR or push a new commit to an open PR which will be published as `<PR number>-<commit full SHA1>`

> 2. Merge a PR to `develop` which will be published as `<PR number>-<commit full SHA1>`

Reviewer: Probably a copy-paste slip. I thought we agreed to use:

```suggestion
2. Merge a PR to `develop` which will be published as `develop-<commit full SHA1>`
```

From a technical point of view, I don't think CircleCI can obtain the PR identifier of the commit that triggered the build on `develop`.

Author: Yeap, copy-paste slip. Fixed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

3. Create a new tag which will be published as `<tag name>`

**How it works:**

CI will run the following commands:

> - `npm install`

Reviewer: ```suggestion
- `npm ci`
```

Reference [this](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3230/files#r591378300) and [this](https://github.com/wordpress-mobile/gutenberg-mobile/compare/mokagio/publish-to-bintray-with-script...mokagio/publish-to-bintray-with-script-demo-2#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47R317).

Author: Fixed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

- `npm run bundle:android`

> - `gutenberg/packages/react-native-bridge/android/publish-aztec-and-bridge.sh <version>`

Reviewer: ```suggestion
- `gutenberg/packages/react-native-bridge/android/publish-aztec-and-bridge.sh <version>`, where `version` is a value computed in CI following the rules described above
```

Or do you think it's too verbose?

Author: Is this really necessary? We already have the main section explaining the version value and `how it works` section is supposed to just give extra insight into what's happening.

Reviewer: Fair enough. Too verbose 👍

---

### Manually deploy a new version of `react-native-bridge`

> _**IMPORTANT:** This should only be used for testing and a version deployed as such shouldn't be merged to WPAndroid's `develop`. This is because we'd bypass the guarantees CI would otherwise give us._

Reviewer: I'm not sure I understand this note.

The following suggestion is the combination of my understanding of the process and of the sentences themselves.

```suggestion
_**IMPORTANT:** This should only be used for testing a version not yet deployed and as such the new version value shouldn't be committed to WPAndroid's `develop`. This is because the process bypasses the guarantees CI would otherwise give us._
```

Author: I didn't fully understand your suggestion either. I re-worked this in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1, let me know what you think.

**How it works:**

Besides checking a few things, `publish-aztec-and-bridge.sh` script will clean the project, copy the bundled JS file into its assets, publish `react-native-aztec` and use that version to build and publish `react-native-bridge`.

---

### Update WPAndroid's binary version

- Find [the version deployed from CI](#deploy-a-new-version-of-react-native-bridge-from-ci) or [the version you've manually deployed](#manually-deploy-a-new-version-of-react-native-bridge)
- Update `ext.gutenbergMobileVersion` property in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle) to the new version

In order to test this, make sure `localGutenbergMobilePath` in your `local-builds.gradle` file is commented out as otherwise the binary version will be ignored.

---

## How to

### How to test a gutenberg-mobile PR in WPAndroid

Assuming that there is no open WPAndroid PR yet:

1. Setup WPAndroid to [work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
2. Switch `gutenberg-mobile` to the branch you're testing
3. Run the app

---

> ### How to share WIP WPAndroid APK

Reviewer: Nitpick

```suggestion
### How to share a WIP WPAndroid APK
```

or 

```suggestion
### How to share WIP WPAndroid APKs
```

Author: Addressed in 21e81ffd7b7a2ea738a9fe0611f82f085f860cf1.

It'll also expect the metro server to be running to fetch the JS bundle.

- [Work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
- [Deploy a new version of `react-native-bridge` from CI](#deploy-a-new-version-of-react-native-bridge-from-ci)
- [Manually deploy a new version of `react-native-bridge`](#manually-deploy-a-new-version-of-react-native-bridge)
- [Update WPAndroid's binary version](#update-wpandroids-binary-version)
- [How to](#how-to)
    - [How to test a gutenberg-mobile PR in WPAndroid](how-to-test-a-gutenberg-mobile-pr-in-wpandroid)
    - [How to share WIP WPAndroid APK](how-to-share-wip-wpandroid-apk)
    - [How to checkout gutenberg-mobile to deployed binary version](#how-to-checkout-gutenberg-mobile-to-deployed-binary-version)

---

### Work with local gutenberg-mobile checkout

- Make sure you run `npm install` or `npm ci`
- Copy [local-builds.gradle-example](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/local-builds.gradle-example) renaming it to `local-builds.gradle`
- Update `localGutenbergMobilePath`to your local `gutenberg-mobile`checkout

> - Run metro server with `npm run start:reset`

Reviewer 2: I find these instructions a bit hard to follow because we don't say _where_ to run each command. Some should be run in `mobile-gutenberg/`, but others happen in the WPAndroid root (i.e., the copying and renaming of `local-builds.gradle-example` which doesn't say where the file should be copied to).

Reviewer 2: One way to help this be clearer might be to group the `npm` steps together and the `WPAndroid` steps together instead of having `npm` -> `WPAndroid` -> `npm`. Just a thought.

Author: I've reworked this in 6a5ca7effc502ea798275d8ced92a54755d14942, let me know if that's a bit clearer.

Author: > One way to help this be clearer might be to group the npm steps together and the WPAndroid steps together instead of having npm -> WPAndroid -> npm. Just a thought.

I actually missed this comment while working on your feedback. So, I am glad I arrived at the same approach :)

Reviewer 2: ![](https://media1.giphy.com/media/7l5al5QCZRMLIPW9Cd/giphy.gif)

```
implementation "$rootProject.gradle.ext.gutenbergMobileBinaryPath:$rootProject.ext.gutenbergMobileVersion"
```

`react-native-bridge` is setup to include all the other projects, so no other action is necessary.

---

### Deploy a new version of `react-native-bridge` from CI

There are a few different ways to do this:

1. Open a new PR or push a new commit to an open PR which will be published as `<PR number>-<commit full SHA1>`
2. Merge a PR to `develop` which will be published as `<develop>-<commit full SHA1>`
3. Create a new tag which will be published as `<tag name>`

**How it works:**

CI will run the following commands:

- `npm ci`
- `npm run bundle:android`
- `gutenberg/packages/react-native-bridge/android/publish-aztec-and-bridge.sh <version>`

---

### Manually deploy a new version of `react-native-bridge`

_**IMPORTANT:** This should only be used for testing.
We shouldn't commit a manually deployed version to WPAndroid's `develop` because the process bypasses the guarantees CI would otherwise give us._

- Make sure you run `npm install` or `npm ci` first
- Run `npm run bundle:android` so the JS bundle is created

> - Run `publish-aztec-and-bridge.sh` script in `react-native-bridge`, giving it a version name such as `locally-built-by-<your_name>-<commit full SHA1>`

Reviewer 2: > giving it a version name

Might be good to be explicit about what "it" is and how you give it a version name, i.e. is this something the script prompts you for or an argument to the script?

Author: I've reworked this in 6a5ca7effc502ea798275d8ced92a54755d14942, let me know if that's clearer.

FYI: The script will also return an example usage if it's not called properly.

Reviewer 2: Looks great! 👍

**How it works:**

Besides checking a few things, `publish-aztec-and-bridge.sh` script will clean the project, copy the bundled JS file into its assets, publish `react-native-aztec` and use that version to build and publish `react-native-bridge`.

---

### Update WPAndroid's binary version

- Find [the version deployed from CI](#deploy-a-new-version-of-react-native-bridge-from-ci) or [the version you've manually deployed](#manually-deploy-a-new-version-of-react-native-bridge)
- Update `ext.gutenbergMobileVersion` property in [build.gradle](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/build.gradle) to the new version

In order to test this, make sure `localGutenbergMobilePath` in your `local-builds.gradle` file is commented out as otherwise the binary version will be ignored.

---

## How to

### How to test a gutenberg-mobile PR in WPAndroid

Assuming that there is no open WPAndroid PR yet:

1. Setup WPAndroid to [work with local gutenberg-mobile checkout](#work-with-local-gutenberg-mobile-checkout)
2. Switch `gutenberg-mobile` to the branch you're testing
3. Run the app

---

### How to share a WIP WPAndroid APK

> 1. Open a `gutenberg-mobile` PR which will [publish a new version](#deploy-a-new-version-of-react-native-bridge-from-ci) or [manually publish a new version](#manually-deploy-a-new-version-of-react-native-bridge)

Reviewer 2: I would rephrase this a bit so there's no chance of someone reading this as saying that opening the PR will "publish a new version or manually publish a new version".
```suggestion
1. Either (a) open a `gutenberg-mobile` PR which will [publish a new version](#deploy-a-new-version-of-react-native-bridge-from-ci) or (b) [manually publish a new version](#manually-deploy-a-new-version-of-react-native-bridge)
```

Author: Addressed in 6a5ca7effc502ea798275d8ced92a54755d14942.

2. Open a WPAndroid PR [updating the binary version](#update-wpandroids-binary-version)
3. Peril will publish a comment to the WPAndroid PR with a link to the APK

Alternatively, once the `react-native-bridge` version is published, we can locally create the APK.

> **Note that there is no way to use composite build to create a WPAndroid APK that can be shared. This is because composite build only works with the local metro server.**

Reviewer 2: Good idea to call this out. 👍

---

### How to checkout gutenberg-mobile to deployed binary version

Whenever the CI publishes `react-native-bridge`, it will either include the full commit hash or the tag in its name.

> So, we can checkout simply checkout the commit has or the tag.

Reviewer 2: ```suggestion
So, we can checkout simply checkout the commit hash or the tag.
```

Author: 🤦 Fixed in 6a5ca7effc502ea798275d8ced92a54755d14942.
